### PR TITLE
Update runbook query for top tenants per ingester

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
Signed-off-by: Dimitar Dimitrov <dimitar.dimitrov@grafana.com>

#### What this PR does

@replay wrote a new query for finding the top tenants on the top loaded ingesters in a cluster. This PR replaces the current two queries that we used to achieve the same during rebalancing of tenants' shards. 

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated
- [x] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by replay** [comment]
I wonder if we could find a way to exclude those tenants which already have a shuffle-shard size that it >= the ingester count... I'll take some more time today to try to figure that out

**Review by pracucci — COMMENTED** [review]
I'm a bit surprised this query works better than the previous one. The new query will return biggest tenants in a pod, but that's not necessarily the contributors to unbalanced series. Unbalanced series may actually be caused by smaller tenants (whose shard size is small).

I would like to have a conversation about this, before changing it, cause I'm not yet convinced the change is correct.

**Comment by replay** [comment]
> The new query will return biggest tenants in a pod, but that's not necessarily the contributors to unbalanced series. Unbalanced series may actually be caused by smaller tenants (whose shard size is small).

This query is based on the following reasoning:

* We want to first find the ingesters which have the most series in the cell, because these are the ingesters where we need to reduce the series count relative to the others
* We want to find the tenants which make up the largest share of active series on these ingesters, because these are the tenants which will have the biggest impact on the total series count of the target ingesters if we spread them out by increasing their shard size

This should allow us to achieve a balanced series count among the ingesters.
If I understand you correctly, then you are aiming for a slightly different goal. Instead of trying to keep the total active series per ingester balanced, you would try to find the tenants that have the most uneven distribution of active series among the ingesters within their shuffle shard, correct?
I think those are two slightly different things whether we aim to a) balance the total active series per ingester or b) the distribution of the active series within each tenant's shuffle shard

**Comment by pracucci** [comment]
> We want to find the tenants which make up the largest share of active series on these ingesters, because these are the tenants which will have the biggest impact on the total series count of the target ingesters if we spread them out by increasing their shard size

I don't fully agree on this. It's not necessarily the tenants having the largest number of series which cause the most impact. I've frequently seen it to be caused by several medium-sized tenants, which may just be hidden by the new query. For example, the 3 biggest tenant in a cluster may just shard across all ingesters, and there's nothing to do with them.

> If I understand you correctly, then you are aiming for a slightly different goal. Instead of trying to keep the total active series per ingester balanced, you would try to find the tenants that have the most uneven distribution of active series among the ingesters within their shuffle shard, correct?

My goal is the same of you: improve the balance of series across ingesters.

> I think those are two slightly different things whether we aim to a) balance the total active series per ingester or b) balance the distribution of the active series within each tenant's shuffle shard

My take is: if you balance the distribution of the active series within each tenant's shard, and you have a reasonably high number of tenants, then you should get a good balance of series across all ingesters.

To understand the previous query, we need to understand how shuffle-sharding is configured. The rationale is explained here:
https://github.com/grafana/mimir/blob/07f5993d099bd624fa191c589f0e95d69a6eacf9/operations/mimir/shuffle-sharding.libsonnet#L27-L32

The idea was targetting shuffle sharding assuming tenants use 50% of their series limit. In this context, outliers are tenants with a number of series close to their limit (at least > 50% of their limit), and that's what the previous query was tying to find.

**Comment by replay** [comment]
> My take is: if you balance the distribution of the active series within each tenant's shard, and you have a reasonably high number of tenants, then you should get a good balance of series across all ingesters.

Interesting, that's a different way of thinking about it than I have applied so far... How would you then handle the following situations:

* you have one or more unbalanced tenants with a shard size that is >= ingester count, then you have no way to rebalance the active series per ingester within that tenant afaik, so you can only try to compensate for the imbalances within these huge tenants by spreading out other tenants that happen to be on the same ingesters where the huge tenants have a high number of series
* your cluster doesn't have a large number of tenants

**Comment by replay** [comment]
> you have one or more unbalanced tenants with a shard size that is >= ingester count, then you have no way to rebalance the active series per ingester within that tenant afaik, so you can only try to compensate for the imbalances within these huge tenants by spreading out other tenants that happen to be on the same ingesters where the huge tenants have a high number of series
> 
> There's nothing you can do anyway for these tenants. There's no way to change the number of series per ingester stored by these tenants (given they already shard across all ingesters), so my take is that you should exclude all of them from your analysis.

Yeah, you can do nothing for these tenants. But if the goal is only to balance the total number of active series per ingester, then you compensate for imbalances in these huge tenants by spreading out smaller tenants accordingly, no?

> In this PR we're talking about unbalanced number of series caused by shuffle sharding (the query in the runbook is under the section "Check if shuffle-sharding shard size is correct"). Unbalanced number of series because of shuffle sharding should be never caused by tenants with a shard size >= ingesters count, because if shard size >= ingesters count, then shuffle sharding is disabled.

Yeah, good point. Maybe this would be the wrong section for this query.

**Comment by replay** [comment]
> Maybe a better compromise could be keeping both queries in the runbook.

I think that would make sense, because depending on the situation one or the other might be more useful

**Comment by dimitarvdimitrov** [comment]
I realized that none of the two queries actually show anything to do with imbalance - the difference for a tenant between the most series they have on any ingester and the least series on any ingester. I put together a query that tries to satisfy the requirements you talked about above:
* only show tenants that have more than 100K series per ingester (on some ingester)
* only show tenants that aren't already sharded to all ingesters
* only show tenants on the ingesters with most series
* show imbalance - the ratio between highest loaded ingester for a tenant vs lowest loaded ingester for a tenant


query looks a bit long, but it was fun to write

```
sort_desc(
    (
        (
            ( # max series per tenant per ingester
                max by (user) (cortex_ingester_memory_series_created_total{namespace="cortex-prod-10"} - cortex_ingester_memory_series_removed_total{namespace="cortex-prod-10"})
                    > 50000 # only show tenants that have more than 100K series per ingester && are over 50% of their per-ingester limits
                    >
                    (
                        0.5
                        *
                        max by(user) (cortex_limits_overrides{namespace="cortex-prod-10",limit_name="max_global_series_per_user"})
                        *
                        scalar(max(cortex_distributor_replication_factor{namespace="cortex-prod-10"}))
                        /
                        count by (user) (cortex_ingester_memory_series_created_total{namespace="cortex-prod-10"})
                    )
            )
            /
            # min series per tenant per ingester
            min by (user) (cortex_ingester_memory_series_created_total{namespace="cortex-prod-10"} - cortex_ingester_memory_series_removed_total{namespace="cortex-prod-10"})
        )
        * on (user) group_left(pod) # get a pod label for the most loaded ingester by the tenant
            group by (user, pod) ( # effectively set it to 1 so that the multuply above (*) is a noop
                topk by (user) (1, cortex_ingester_memory_series_created_total{namespace="cortex-prod-10"} - cortex_ingester_memory_series_removed_total{namespace="cortex-prod-10"})
            )
    )
    and on(user) ( # intersection with the tenants which don't have series on all ingesters
        count by (user) (cortex_ingester_memory_series_created_total{namespace="cortex-prod-10"})
        !=
        scalar(count(count by (pod) (cortex_ingester_memory_series_created_total{namespace="cortex-prod-10"})))
    )
    # and on(pod) ( # intersection with the 10 most loaded ingesters
    #     topk(10, sum by (pod) (cortex_ingester_memory_series{namespace="cortex-prod-10"}))
    # )
)
```

I am unsure if we need to only show "most overloaded ingesters." I've left that condition commented out near the bottom of the query. If we are aiming for the most unbalanced tenants, those can live in many different ingesters. In fact we get some tenants with imbalance ratio of 7.8 in one of our cells that aren't on the top ingesters.

I cross-checked the tenants with the query below to show the series per ingester

```
sum by (user, pod) (cortex_ingester_memory_series_created_total{user="383720", namespace="cortex-prod-10"} - cortex_ingester_memory_series_removed_total{namespace="cortex-prod-10"})
```

Do you guys think that this query can replace the existing one and the one proposed in the PR?

**Comment by dimitarvdimitrov** [comment]
let me know when you've had a chance to take a look at the above.

The main issues I was trying to solve is having to run more than one query and having to manually estimate is a tenant should/can be resharded. If that query is too complex for the job or you think it's inaccurate I'm ok to just include both queries you talked about and get this merged. I'm also more than happy to incorporate any feedback about it.

**Comment by replay** [comment]
I'm not sure I fully understand the query...
What does this part do?
```
                    (
                        0.5
                        *
                        max by(user) (cortex_limits_overrides{namespace="cortex-prod-10",limit_name="max_global_series_per_user"})
                        *
                        scalar(max(cortex_distributor_replication_factor{namespace="cortex-prod-10"}))
                        /
                        count by (user) (cortex_ingester_memory_series_created_total{namespace="cortex-prod-10"})
                    )
```
Should this be getting the tenants where the series count is above 50% of the limit? Shouldn't it be:
```
                    (
                        0.5
                        *
                        max by(user) (cortex_limits_overrides{namespace="cortex-prod-10",limit_name="max_global_series_per_user"})
                        /
                        scalar(max(cortex_distributor_replication_factor{namespace="cortex-prod-10"}))
                        *
                        count by (user) (cortex_ingester_memory_series_created_total{namespace="cortex-prod-10"})
                    )
```

I'm sure it's useful to find the most unbalanced tenants in many scenarios, but I think in situations where one or both of these conditions are true:

* The total tenant count is low
* Few tenants are much bigger than all the others

there is still benefit in being able to look up the tenants which have the most active series on the ingesters with the most active series in order to spread them out. This is what your original PR does, so I don't think this query can fully replace it if I understand it correctly. 

I think the query in your original PR would be even more useful if we would filter out the tenants who's shuffle-shard already includes all ingesters.

**Comment by dimitarvdimitrov** [comment]
thank you for the feedback


> ```
>                     (
>                         0.5
>                         *
>                         max by(user) (cortex_limits_overrides{namespace="cortex-prod-10",limit_name="max_global_series_per_user"})
>                         *
>                         scalar(max(cortex_distributor_replication_factor{namespace="cortex-prod-10"}))
>                         /
>                         count by (user) (cortex_ingester_memory_series_created_total{namespace="cortex-prod-10"})
>                     )
> ```

This shows tenants which are 50% over their per-instance limit on at least one ingester. I guess I misunderstood how this was working in the original query in the runbook (the one this PR replaces as of now).

The query in my comment uses the ratio between max series per ingester in the shard vs the min series per ingester in the shard (series = series for this tenant, not for the whole ingester). _Would it be better to instead show the difference between them?_ And maybe filter on the difference. If the difference is significant (say, 20K series or more) then we display the tenant and the ingester.

File: docs/sources/operators-guide/mimir-runbooks/_index.md [hunk 1]
```diff
@@ -70,32 +70,35 @@ How to **fix** it:
 2. **Check if shuffle-sharding shard size is correct**<br />
 
 - When shuffle-sharding is enabled, we target up to 100K series / tenant / ingester assuming tenants on average use 50% of their max series limit.
-- Run the following **instant query** to find tenants that might cause higher pressure on some ingesters:
+- Run the following **instant query** to find tenants that might cause higher pressure on ingesters. The query excludes tenants which are already sharded across all tenants:
```
Comment (replay) [inline docs/sources/operators-guide/mimir-runbooks/_index.md:73]:
```suggestion
- Run the following **instant query** to find tenants that might cause higher pressure on ingesters. The query excludes tenants which are already sharded across all ingesters:
```

File: docs/sources/operators-guide/mimir-runbooks/_index.md [hunk 2]
```diff
@@ -70,32 +70,35 @@ How to **fix** it:
...
-    sum by(user) (cortex_ingester_memory_series_created_total{namespace="<namespace>"}
-    -
-    cortex_ingester_memory_series_removed_total{namespace="<namespace>"})
+  topk by (pod) (
+    5, # top 5 users per ingester
+    (
+      sum by (user, pod) (
+        cortex_ingester_memory_series_created_total{namespace="<namespace>"} - cortex_ingester_memory_series_removed_total{namespace="<namespace>"}
+      )
+      > on (user) group_left() # show tenants that are exeeding 50% of their series limits added acorss ingesters
```
Comment (replay) [inline docs/sources/operators-guide/mimir-runbooks/_index.md:82]:
```suggestion
      > on (user) group_left() # show tenants that are exceeding 50% of their series limits added across ingesters
```

File: docs/sources/operators-guide/mimir-runbooks/_index.md [hunk 3]
```diff
@@ -70,32 +70,35 @@ How to **fix** it:
...
+      (
+        max by(user) (cortex_limits_overrides{namespace="<namespace>",limit_name="max_global_series_per_user"})
+        *
+        scalar(max(cortex_distributor_replication_factor{namespace="<namespace>"}))
+        *
+        0.5
+      )
+      > 200000
```
Comment (replay) [inline docs/sources/operators-guide/mimir-runbooks/_index.md:90]: where does the `200k` come from? is the idea to make sure that only tenants that have a minimum size get shown?

Comment (dimitarvdimitrov) [reply docs/sources/operators-guide/mimir-runbooks/_index.md]: I took the 200k from the previous query, I think your understanding is correct

**Review by pracucci — APPROVED** [review]
I haven't run the query myself to test it, but the reasoning behind the proposed query makes sense to me. Let's give it a try!

**Pull request merged**